test/functional/test_framework/util.py

```diff
@@ -267,6 +267,13 @@ def satoshi_round(amount: Union[int, float, str], *, rounding: str) -> Decimal:
     """Rounds a Decimal amount to the nearest satoshi using the specified rounding mode."""
     return Decimal(amount).quantize(SATOSHI_PRECISION, rounding=rounding)
 
+def ensure_for_helper_internal(*, duration, predicate):
+    time_end = time.time() + duration
+    predicate_source = "''''\n" + inspect.getsource(predicate) + "'''"
+    while time.time() < time_end:
+        if not predicate():
+            raise AssertionError("Predicate {} became false within {} seconds".format(predicate_source, duration))
+        time.sleep(0.05)
```

Comment:
Isn't it a bit aggressive to execute the provided predicate every 0.05 seconds for the entire duration?
Beyond flooding the logs, the predicate likely calls an RPC command that locks `cs_main` or some other important mutex, which slows down the node.

For instance, the `feature_assumeutxo.py` one, will call `getpeerinfo()` 60 times.

Reply:
I simply took the same interval as `wait_until` uses. If the slowdown was an issue it should be much more apparent from the use there, so I did a simple benchmark and increasing the sleep from `0.05` to `0.2` lead to a considerable slowdown of the test suite overall (8min to 10:45min), so I don't think this would be an issue here either.

We could still lower the interval. Since we generally don't expect the predicate to become false it shouldn't hurt performance as much as changing it in `wait_until`. It means we would have fewer logs and if a failure occurs it would take a bit longer to detect it.

Reply:
> And, in fact, we could implement it using the wait_until function too

Yes, I tried this and decided against it, see the description of the PR.

> I’d suggest simply waiting for the entire duration without intermediate checks.

I have also considered this and decided against it because I don't think it's better. I have also described this in the PR description already. Since the polling doesn't seem to lead to a real performance issue I think it's better to fail faster if we fail. This also uncovers potential races, as mentioned in the description.

Reply:
> > And, in fact, we could implement it using the wait_until function too
> 
> Yes, I tried this and decided against it, see the description of the PR.

Ah, missed it. Sorry.

> > I’d suggest simply waiting for the entire duration without intermediate checks.
> 
> I have also considered this and decided against it because I don't think it's better. I have also described this in the PR description already. Since the polling doesn't seem to lead to a real performance issue I think it's better to fail faster if we fail. This also uncovers potential races, as mentioned in the description.

hmm, so you are thinking about a condition that passes at the beginning, does not passes somewhere during the test sleep time and ends up passing at the end. Maybe the test condition should be changed if such scenario exist?

I'm not sure we can really ensure that this won't lead to performance issues. It's abnormal to fail here, as it's something we don't expect during the test. E.g. we would be making 59 additional RPC calls in `feature_assumeutxo.py`, which would flood the active logs (logs we may need if the test fails later for some other reason after wise), all while looking for a failure that will rarely occur.
I mean, for current use cases, it probably doesn't matter. But could this be an issue for future uses? It seems to be something it could be misused very easily.

Reply:
> hmm, so you are thinking about a condition that passes at the beginning, does not passes somewhere during the test sleep time and ends up passing at the end. Maybe the test condition should be changed if such scenario exist?

Yes, exactly, and in order to know that it should be changed the person writing the test will see that `ensure_for` fails. This is the protection against introducing flakiness that I mentioned.

> which would flood the active logs

I don't know if that's a problem because I always only grep the logs anyway.

> I'm not sure we can really ensure that this won't lead to performance issues.

The benchmarks say no but if you have another way to show that there is a performance impact. Maybe that's not a guarantee but if there is any it appears to be very small.

However, I have made the delay configurable now so this can be adjusted and I have set the default to 0.2 instead of 0.05. That should keep impact on the logs (and performance) small. I hope that works for everyone.